flang-rt/lib/runtime/exceptions.cpp

```diff
@@ -76,22 +76,69 @@ uint32_t RTNAME(MapException)(uint32_t excepts) {
   return except_value;
 }
 
+// The following exception processing routines have a libm call component,
+// and where available, an additional component for handling the nonstandard
+// ieee_denorm exception. The denorm component does not subsume the libm
+// component; both are needed.
+
+void RTNAME(feclearexcept)(uint32_t excepts) {
+  feclearexcept(excepts);
+#if defined(_MM_EXCEPT_DENORM)
+  _mm_setcsr(_mm_getcsr() & ~(excepts & _MM_EXCEPT_MASK));
+#endif
+}
+void RTNAME(feraiseexcept)(uint32_t excepts) {
+  feraiseexcept(excepts);
+#if defined(_MM_EXCEPT_DENORM)
+  _mm_setcsr(_mm_getcsr() | (excepts & _MM_EXCEPT_MASK));
+#endif
+}
+uint32_t RTNAME(fetestexcept)(uint32_t excepts) {
+#if defined(_MM_EXCEPT_DENORM)
+  return (_mm_getcsr() & _MM_EXCEPT_MASK & excepts) | fetestexcept(excepts);
+#else
+  return fetestexcept(excepts);
+#endif
+}
+void RTNAME(fedisableexcept)(uint32_t excepts) {
+  extern int fedisableexcept(int);
+  fedisableexcept(excepts);
+#if defined(_MM_EXCEPT_DENORM)
+  _mm_setcsr(_mm_getcsr() | ((excepts & _MM_EXCEPT_MASK) << 7));
+#endif
+}
+void RTNAME(feenableexcept)(uint32_t excepts) {
+  extern int feenableexcept(int);
+  feenableexcept(excepts);
+#if defined(_MM_EXCEPT_DENORM)
+  _mm_setcsr(_mm_getcsr() & ~((excepts & _MM_EXCEPT_MASK) << 7));
+#endif
+}
+uint32_t RTNAME(fegetexcept)() {
+  extern int fegetexcept(void);
+#if defined(_MM_EXCEPT_DENORM)
+  return (63 - ((_mm_getcsr() >> 7) & _MM_EXCEPT_MASK)) | fegetexcept();
+#else
+  return fegetexcept();
+#endif
+}
+
 // Check if the processor has the ability to control whether to halt or
 // continue execution when a given exception is raised.
 bool RTNAME(SupportHalting)([[maybe_unused]] uint32_t except) {
 #ifdef __USE_GNU
```

Comment:
Do you need to move the `#ifdef __USE_GNU` to the new helper definitions?

Reply:
Done; thanks